File: pyfenstein3d/engine/door.py
import math
from .block import Block
from .vector2d import Vector2d
from .config import DOOR_OPEN_VELOCITY
from .config import DOOR_INTERVAL_CLOSE
from .animation import Animation

class Door(Block):
    def __init__(self, x: float, y: float, type_id: int, is_vertical: bool):
        super().__init__(x, y, type_id, True)
        self.__is_vertical = is_vertical
        self._is_solid = True
        self.__is_opened = False
        self.__is_opening = False
        self.__is_closing = False
        self.__original_x = x
        self.__original_y = y
        self.__closing_interval = 0
        self.__animations = {
            "open": Animation(DOOR_OPEN_VELOCITY, self.__on_opening_animate, self.__on_opening_animate_end),
            "wait": Animation(DOOR_INTERVAL_CLOSE, on_animate_end=self.__on_waiting_animate_end),
            "close": Animation(DOOR_OPEN_VELOCITY, self.__on_closing_animate, self.__on_closing_animate_end)
        }


    @property
    def is_vertical(self):
        return self.__is_vertical

    def __open(self):
        self.__animations["open"].start()

    def __close(self):
        self._is_solid = True
        self.__animations["close"].start()

    def __on_opening_animate(self, factor):
        vector = Vector2d(factor if not self.is_vertical else 0, factor if self.is_vertical else 0)
        self._vector2d = Vector2d(self.__original_x, self.__original_y) - vector

    def __on_opening_animate_end(self):
        self.__is_opening = False
        self.__is_opened = True
        self._is_solid = False
        self.__animations["wait"].start()

    def __on_waiting_animate_end(self):
        self.__close()

    def __on_closing_animate(self, factor):
        vector = Vector2d(1 - factor if not self.is_vertical else 0, 1 - factor if self.is_vertical else 0)
        self._vector2d = Vector2d(self.__original_x, self.__original_y) - vector

    def __on_closing_animate_end(self):
        self.__is_closing = False
        self.__is_opened = False


    def _get_bounds(self, fov_ang: float):
        bounds = []
        bounds.append(self._vector2d)
        bounds.append(self._vector2d + Vector2d(1 if not self.__is_vertical else 0, 1 if self.__is_vertical else 0))
        return bounds

    def update(self, delta_time: float, persons: []):
        for anim in self.__animations.values():
            anim.update(delta_time)
                
    def interacted(self):
        if not self.__is_opening and not self.__is_closing:
            if not self.__is_opened:
                self.__open()

File: engine/door.go
package engine

// Door is a block that can open and close.
type Door struct {
	block
	isVertical   bool
	isOpened     bool
	isOpening    bool
	isClosing    bool
	originalX    float64
	originalY    float64
	openAnim     *Animation
	waitAnim     *Animation
	closeAnim    *Animation
}

// NewDoor creates a new Door at the given position.
func NewDoor(x, y float64, typeID int, isVertical bool) *Door {
	d := &Door{
		block:      newBlock(x, y, typeID, true, false),
		isVertical: isVertical,
		originalX:  x,
		originalY:  y,
	}
	d.openAnim = NewAnimation(DoorOpenVelocity, d.onOpeningAnimate, d.onOpeningAnimateEnd)
	d.waitAnim = NewAnimation(DoorIntervalClose, nil, d.onWaitingAnimateEnd)
	d.closeAnim = NewAnimation(DoorOpenVelocity, d.onClosingAnimate, d.onClosingAnimateEnd)
	return d
}

func (d *Door) IsVertical() bool { return d.isVertical }

func (d *Door) onOpeningAnimate(factor float64) {
	var vx, vy float64
	if !d.isVertical {
		vx = factor
	} else {
		vy = factor
	}
	d.setXY(d.originalX-vx, d.originalY-vy)
}

func (d *Door) onOpeningAnimateEnd() {
	d.isOpening = false
	d.isOpened = true
	d.isSolid = false
	d.waitAnim.Start(0)
}

func (d *Door) onWaitingAnimateEnd() {
	d.doClose()
}

func (d *Door) onClosingAnimate(factor float64) {
	var vx, vy float64
	if !d.isVertical {
		vx = 1 - factor
	} else {
		vy = 1 - factor
	}
	d.setXY(d.originalX-vx, d.originalY-vy)
}

func (d *Door) onClosingAnimateEnd() {
	d.isClosing = false
	d.isOpened = false
}

func (d *Door) doOpen() {
	d.openAnim.Start(0)
}

func (d *Door) doClose() {
	d.isSolid = true
	d.closeAnim.Start(0)
}

func (d *Door) GetBounds(fovAng float64) []*Vector2d {
	var dx, dy float64
	if !d.isVertical {
		dx = 1
	} else {
		dy = 1
	}
	return []*Vector2d{
		d.pos,
		d.pos.Add(NewVector2d(dx, dy)),
	}
}

func (d *Door) IsInFov(pos *Vector2d, fovAng, fovDelta, dist float64) bool {
	return blockIsInFov(d.GetBounds(fovAng), pos, fovAng, fovDelta, dist)
}

// Update advances all door animations by deltaTime.
func (d *Door) Update(deltaTime float64, persons []Blocker) {
	d.openAnim.Update(deltaTime)
	d.waitAnim.Update(deltaTime)
	d.closeAnim.Update(deltaTime)
}

// Interacted toggles the door open/close when the player interacts.
func (d *Door) Interacted() {
	if !d.isOpening && !d.isClosing {
		if !d.isOpened {
			d.doOpen()
		}
	}
}
